File: euler_046.py
from time import time
from primePy import primes
from math import sqrt

start = time()

limit = 10000

primenums = set(primes.upto(limit))
twas = set(2*n*n for n in range(int(sqrt(limit/2) + 1)))
odd_comp = set(2 * k + 1 for k in range(int(limit/2) + 1) if not 2 * k - 1 in primenums)

sums = set()

for p in primenums:
    for t in twas:
        sums.add(p + t)

print(set(o for o in odd_comp if o not in sums))


print("Composite number: " + str(0))

stop = time()
print("Time: " + str(stop-start))

File: euler_046.js
const start = Date.now();

// Primes
const upperBound = 10000;

let sieve = Array(upperBound);
sieve.fill(true);
sieve[0] = false;
sieve[1] = false;

let primes = [];

for (let c = 2;c < upperBound; c++) {
    if (sieve[c]) {
        primes.push(c);
        let j = 2;
        while (c * j < upperBound) {
            sieve[c * j] = false;
            j++;
        }
    }
}



// 2*m^2s
let twiceASquare = Array(upperBound);
twiceASquare.fill(false);
let i = 1;

while (2 * i * i < upperBound) {
    twiceASquare[2 * i * i] = true;
    i++;
}

let stillLooking = true;
let res;

for (let c = 9; c < upperBound && stillLooking; c += 2) {
    if (!sieve[c]) {
        let count = 0;
        for (const p of primes.filter(x => x < c)) {
            if (twiceASquare[c - p]) {
                count++;
            }
        }
        if (count === 0) {
            stillLooking = false;
            res = c;
        }
    }
}


const end = Date.now();

console.log("Execute time: ", end - start, " The smallest odd composite number is: ", res);